### PR TITLE
Drake/feature/uart interface

### PR DESCRIPTION
**Pull request by DrakBoul**
Created UART interface as well as extremely simple test program to confirm that the interface works using an arduino uno. Implemented methods reading and writing message structures as well as raw bytes. 

The serialport Crate relies on libudev which I've installed locally on my own machine to get the library and test program to properly compile and run. But our build git workflow fails because of this dependency. Need to look Further into this before merging but as of now the UART interface is complete. 

Please give feedback on how we can potentially fix the build issues as well as any other features we want to add to the interface.

### CODE REVIEW

**Comment by rrasmuss4200** [comment]
Looks great. Maybe we can test these on Saturday and then merge both I2C and UART

File: ex3_shared_libs/interfaces/uart/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,59 @@
+use message_structure::{deserialize_msg, serialize_msg, Msg};
+use serialport::TTYPort;
+use std::io::{Error as IoError, Read, Write};
+
+pub trait Interface {
+    fn read(&mut self) -> Result<Msg, IoError>;
+    fn send(&mut self, msg: &Msg) -> Result<usize, IoError>;
```
Comment (rcunrau) [inline ex3_shared_libs/interfaces/uart/src/lib.rs:8]: Not meaning to pick, but usually you use read with write and send with recv :-)

File: ex3_shared_libs/interfaces/uart/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+pub struct UARTInterface {
+    interface: TTYPort,
+    device_name: String,
+    baud_rate: u32,
```
Comment (rcunrau) [inline ex3_shared_libs/interfaces/uart/src/lib.rs:15]: Do we have to be concerned about things like parity or stop bits?

**Review by rcunrau — APPROVED** [review]

File: ex3_shared_libs/interfaces/uart/src/lib.rs (hunk 2 above)
Comment (DrakBoul) [reply ex3_shared_libs/interfaces/uart/src/lib.rs:15]: They are additional fields that are associated with the SerialPortBuilder that can set parity and stop bit parameters. I was planning on adding this in near future.

File: ex3_shared_libs/interfaces/uart/src/lib.rs (hunk 1 above)
Comment (DrakBoul) [reply ex3_shared_libs/interfaces/uart/src/lib.rs:8]: Got it. The purpose for these naming conventions was mainly due to how the interface trait was defined in the tcp lib. As of right now we are re-defining this trait in each interface lib (besides ipc). We could do with refactor that defines this trait a single time in a separate file, at this point we can switch the naming conventions.

**Comment by DrakBoul** [comment]
> Looks great. Maybe we can test these on Saturday and then merge both I2C and UART

As pointed out by Ron this interface needs a bit more work with configuring port parameters. I also want to add some functionality to clear the input and output buffer of the device. Probably not going to be done by work session. I2C should be good to merge however as it was already tested.

File: ex3_shared_libs/interfaces/uart/src/lib.rs (hunk 2 above)
Comment (DrakBoul) [reply ex3_shared_libs/interfaces/uart/src/lib.rs]: created an additional constructor that allows the user to open the port using a SerialPortBuilder object. Using this struct they can modify parameters like control flow, data bits, stop bits, parity, as well as a timeout on there operations.

**Review by rrasmuss4200 — APPROVED** [review]
Saw this working on Saturday. Looking forward to testing it when/if we get the DFGM board

**Pull request merged**